### PR TITLE
store terraform state

### PR DESCRIPTION
**Pull request by dexhorthy**
What I Did
------------

Store terraform state after terraform apply runs. Restore it before "init".

How I Did it
------------

- new function `persistState` used by `Daemonless` and `Fork` terraformers.
(spolier alert, they both fork)
- vendor in subset of terraform code. Vendoring in the whole thing is still
tricky, but I was able to get the types loaded in, which lets us use that
library for unmarshalling the state.
- store state after "apply" runs

How to verify it
------------

Run a ship spec with terraform in it, inspect state.json afterward

Description for the Changelog
------------
Store state from `terraform` in ship state.json

:ship:

### CODE REVIEW

File: integration/base/terraform/expected/.ship/release.yml [hunk 1]
```diff
@@ -3,13 +3,15 @@ assets:
     - terraform:
         dest: ./terraform/random.tf
         inline: |
-          resource "random_id" "server" {
-            byte_length = {{repl ConfigOption "id_length" | ParseInt}}
+          resource "local_file" "foo" {
+            content     = "foo!"
+            filename = "/tmp/foo.bar"
```
Comment (laverya) [inline integration/base/terraform/expected/.ship/release.yml]: do we no longer support template functions here?

File: pkg/lifecycle/terraform/daemonless.go [hunk 2]
```diff
@@ -58,18 +68,26 @@ func (t *DaemonlessTerraformer) WithStatusReceiver(
 		Terraform:     t.Terraform,
 		Viper:         t.Viper,
 		FS:            t.FS,
+		StateManager:  t.StateManager,
+		StateSaver:    t.StateSaver,
```
Comment (dexhorthy) [inline pkg/lifecycle/terraform/daemonless.go]: need a restorer here

File: pkg/lifecycle/terraform/daemonless.go [hunk 3]
```diff
@@ -161,7 +185,7 @@ func (t *DaemonlessTerraformer) plan(dir string) (string, bool, error) {
 
 	// we really shouldn't write plan to a file, but this will do for now
 	cmd := t.Terraform(dir)
-	cmd.Args = append(cmd.Args, "plan", "-input=false", "-out=plan")
+	cmd.Args = append(cmd.Args, "plan", "-input=false", "-out=plan.tfplan")
```
Comment (dexhorthy) [inline pkg/lifecycle/terraform/daemonless.go]: added this extension

File: pkg/lifecycle/terraform/terraformer.go [hunk 4]
```diff
@@ -64,19 +72,28 @@ func (t *ForkTerraformer) WithStatusReceiver(status daemontypes.StatusReceiver)
 		Terraform:     t.Terraform,
 		Viper:         t.Viper,
 		FS:            t.FS,
+		StateManager:  t.StateManager,
```
Comment (dexhorthy) [inline pkg/lifecycle/terraform/terraformer.go]: this needs persistState/restoreState

File: pkg/lifecycle/terraform/terraformer.go [hunk 5]
```diff
@@ -135,7 +155,12 @@ func (t *ForkTerraformer) Execute(ctx context.Context, release api.Release, step
 		<-t.Daemon.MessageConfirmedChan()
 	}
 
-	return nil
+	err = t.StateSaver(debug, t.FS, t.StateManager, dir)
+	if err != nil {
+		return errors.Wrapf(err, "persist terraform state")
+	}
+
+	return err
```
Comment (dexhorthy) [inline pkg/lifecycle/terraform/terraformer.go]: this should be nil

Comment (laverya) [reply pkg/lifecycle/terraform/terraformer.go]: Either make this a `return nil` or `return errors.Wrapf(err, "persist terraform state")` and get rid of the conditional

```suggestion
	return errors.Wrapf(err, "persist terraform state")
```

`errors.Wrap(nil, "string) == nil`

File: integration/base/terraform/expected/.ship/release.yml (hunk 1 above)
Comment (dexhorthy) [reply integration/base/terraform/expected/.ship/release.yml]: we do, I was just trying to make these more reproducible. I can actually roll some of that back now that we're restoring the state

**Pull request merged**